Add the Workspace::purge_all_caches method

emilyalbini opened the pull request:
This PR adds a method to purge the contents of all the caches, with the goal of freeing up disk space.

emilyalbini committed:
buildtest: add a dedicate crate for process-lines

This will speed up running tests, as it will be possible to run
process-lines in parallel with hello-world.

emilyalbini committed:
workspace: add Workspace::purge_all_caches method

Kixiron commented on src/workspace.rs:
```diff
@@ -220,6 +220,31 @@ impl Workspace {
+            self.cargo_home().join("registry").join("cache"),
+        ];
+
+        for index in std::fs::read_dir(self.cargo_home().join("registry").join("index"))? {
+            let index = index?;
+            if index.file_type()?.is_dir() {
+                paths.push(index.path().join(".cache"));
+            }
+        }
```
Is there a reason to manually chain things instead of just running `remove_dir_all` on the top-level dirs?

emilyalbini replied:
Yes, we want to keep the git repository of each index, but we want to remove the `.cache` directory inside each index.

Nemo157 commented on tests/integration/purge_caches.rs:
```diff
@@ -0,0 +1,103 @@
+
+    // Simulate a build, which is going to fill up the caches.
+    for krate in &crates {
+        krate.fetch(&workspace)?;
```
Could this cause an index update if it has changed between the init and now?

emilyalbini replied:
No, rustwide doesn't use `cargo` to download crates or git repositories.

emilyalbini replied:
Whelp there was a race condition anyway.

Nemo157 replied:
Ah, yes, but what about the `check` call during the build?

```
file .workspaces/purge-caches/cargo-home/registry/index/github.com-1ecc6299db9ec823/.git/FETCH_HEAD changed
```

https://github.com/rust-lang/rustwide/pull/26/checks?check_run_id=735677032#step:6:528

Nemo157 replied:
(Though I must still be missing something cause it looks like the registry should be readonly during that).

emilyalbini replied:
The index is updated while generating the lockfile for `lazy_static`. I'll address this tomorrow.

emilyalbini committed:
tests: ignore the index while testing caches purging

emilyalbini committed:
tests: check file size instead of content sha in purge_caches

Hashing all the files in the workspace was taking a lot of time (around
40 seconds on my computer, with a NVMe disk), significantly slowing down
the amount of time needed for running the tests.

This commit switches to comparing file sizes, as after all we only care
about the workspace size not increasing when purging all caches.

emilyalbini commented:
@Nemo157 fixed the check detecting index updates.

In the meantime I also changed the way changes are found: comparing the SHAs took a huge amount of time (around 40 seconds on my NVMe), slowing down running `cargo test` a lot. Since we only care about sizes not increasing when purging the caches, the check now compares file sizes instead of file contents.

Nemo157 commented:
If you applied the `should_ignore` as a `filter_entry` during the `walkdir` then you could avoid collecting details on those files (given that the index is likely to be the major component this would save a lot of time).

emilyalbini committed:
tests: apply should_ignore beforehand in purge_caches tests

emilyalbini commented:
@Nemo157 good catch! Done.
